# hemoglobin-simulation
### A Python project that models the biochemical processes of transcription, translation, and mutations

This project develops simulations of biological processes (transcription, translation, and mutations) to predict a proteinâ€™s 3D structure from the DNA sequence of the corresponding gene. The FASTA file for a DNA sequence is opened, and the amino acid sequence of the corresponding protein is predicted. Different types of DNA mutations are modeled (point mutations, insertions, and deletions), and plots of sequence homology are constructed to compare the predicted mutated protein to the non-mutated protein. Torsion angles within the protein are also calculated, and a Ramachandran plot is constructed in order to visualize the secondary structures present in the 3D folded protein.

### FASTA File Parsing
A FASTA file for any gene sequence can be opened here. The file is parsed line by line, and the nucleotide chain is stored in the variable "sequence".

### Transcription and Translation
The transcription() function returns the complementary mRNA chain for the DNA sequence of interest according to the complementary base pairing rules. The translation() function simulates the standard genetic code and returns the predicted amino acid sequence to be printed.

### Mutations
The point_mutation() function changes the base of one random nucleotide in a random position in the DNA sequence. The insertion() function inserts one random nucleotide after a random position in the sequence. The deletion() function deletes one nucleotide from a random position in the sequence. The subsequent simulations run trials of these random mutations on the DNA sequence, and then transcription and translation are simulated to generate a mutated protein sequence. The mutated sequence is compared to the original non-mutated sequence using percent similarity. The summary of all trials is shown in a histogram.

### Torsion Angles
The BioPython package is used to calculate all of the torsion angles in a protein sequence based on the protein's PDB file. A Ramachandran plot is generated by plotting all torsion angles in the form (&phi;,&psi;). The clustering of these points provides insight into the types and relative quantities of the secondary structures present in the protein.
