{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b30acc62",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "import random as r\n",
    "import matplotlib.pyplot as plt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "909dd50d",
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open('HBB_FASTA.txt')  #open a FASTA file with the gene sequence\n",
    "sequence = ''\n",
    "for line in f:\n",
    "    s = line.split()\n",
    "    if line[0] == '>': #ignore lines that do not contain gene sequences\n",
    "        continue\n",
    "    else:\n",
    "        sequence = sequence + s[0]  #transfer the data from the file to the string \"sequence\"\n",
    "sequence = sequence.replace(' ','')  \n",
    "sequence = sequence.replace('\\n','')\n",
    "# makes sure there are no spaces in the string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f90b5b6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Transcription\n",
    "def transcription(mutated_sequence):\n",
    "    template_sequence = ''\n",
    "    transcription_sequence = ''\n",
    "    mutated_sequence = mutated_sequence[mutated_sequence.find('ATG'):]\n",
    "    #translation starts when the ribosome encounters the start codon 'AUG' in \n",
    "    #the mRNA codon.  This corresponds to 'TAC' in the complementary strand, \n",
    "    #which corresponds to 'ATG' in the normal coding sequence.  As a result,\n",
    "    #any noncoding nucleotides in the coding sequence before the start codon will\n",
    "    #be discarded\n",
    "    \n",
    "    for i in mutated_sequence:\n",
    "        if i == 'A':\n",
    "            template_sequence += 'T'\n",
    "        elif i == 'T':\n",
    "            template_sequence += 'A'\n",
    "        elif i == 'G':\n",
    "            template_sequence += 'C'\n",
    "        elif i == 'C':\n",
    "            template_sequence += 'G'\n",
    "        #Transcription occurs on the complementary strand to the normal DNA coding \n",
    "        #strand, so each DNA nucleotide is replaced with its DNA complement to \n",
    "        #generate the complementary template strand\n",
    "       \n",
    "    for i in template_sequence:\n",
    "        if i == 'A':\n",
    "            transcription_sequence += 'U'\n",
    "        elif i == 'T':\n",
    "            transcription_sequence += 'A'\n",
    "        elif i == 'G':\n",
    "            transcription_sequence += 'C'\n",
    "        elif i == 'C':\n",
    "            transcription_sequence += 'G'\n",
    "        #These replace each DNA nucleotide with the complementary mRNA nucleotide\n",
    "        #and add it to the new transcription sequence\n",
    "    return transcription_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "bf30205b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Translation\n",
    "def translation(transcription_sequence):\n",
    "    \n",
    "    codons = []\n",
    "    protein = ''\n",
    "    for i in range(int(len(transcription_sequence)/3)):  \n",
    "        codons.append(transcription_sequence[:3])  #append each set of 3 amino acids to a list of codons\n",
    "        transcription_sequence = transcription_sequence[3:]  #move to the next codon\n",
    "\n",
    "    count=0\n",
    "    for i in codons:\n",
    "        if i == 'AUG': \n",
    "            codons = codons[count:] \n",
    "            break\n",
    "        count+=1\n",
    "        #coding DNA starts at the start codon 'AUG', so this \n",
    "        #ensure that all noncoding nucleotides before this point is discarded\n",
    "\n",
    "    for i in codons:\n",
    "        if i == 'UUU' or i == 'UUC':\n",
    "            protein +='F'\n",
    "        elif i == 'UUA' or i=='UUG' or i=='CUU' or i=='CUA' or i=='CUG' or i=='CUC':\n",
    "            protein +='L'\n",
    "        elif i == 'AUU' or i=='AUC' or i=='AUA':\n",
    "            protein +='I'\n",
    "        elif i == 'AUG':\n",
    "            protein +='M'\n",
    "        elif i == 'GUU' or i=='GUC' or i=='GUA' or i=='GUG':\n",
    "            protein +='V'\n",
    "        elif i == 'UCU' or i=='UCC' or i=='UCA' or i=='UCG' or i=='AGU' or i=='AGC':\n",
    "            protein +='S'\n",
    "        elif i== 'CCU' or i=='CCC' or i=='CCA' or i=='CCG':\n",
    "            protein +='P'\n",
    "        elif i=='ACU' or i=='ACC' or i=='ACA' or i=='ACG':\n",
    "            protein +='T'\n",
    "        elif i=='GCU' or i=='GCC' or i=='GCA' or i=='GCG':\n",
    "            protein +='A'\n",
    "        elif i=='UAU' or i=='UAC':\n",
    "            protein +='Y'\n",
    "        elif i=='CAU' or i=='CAC':\n",
    "            protein +='H'\n",
    "        elif i=='CAA' or i=='CAG':\n",
    "            protein +='Q'\n",
    "        elif i=='AAU' or i=='AAC':\n",
    "            protein +='N'\n",
    "        elif i=='AAA' or i=='AAG':\n",
    "            protein +='K'\n",
    "        elif i=='GAU' or i=='GAC':\n",
    "            protein +='D'\n",
    "        elif i=='GAA' or i=='GAG':\n",
    "            protein +='E'\n",
    "        elif i=='UGU' or i=='UGC':\n",
    "            protein +='C'\n",
    "        elif i=='UGG':\n",
    "            protein +='W'\n",
    "        elif i=='CGU' or i=='CGC' or i=='CGA' or i=='CGG' or i=='AGA' or i=='AGG':\n",
    "            protein +='R'\n",
    "        elif i=='GGU' or i=='GGC' or i=='GGA' or i=='GGG':\n",
    "            protein +='G'\n",
    "        elif i=='UAA' or i=='UAG' or i=='UGA':\n",
    "            break\n",
    "        else:\n",
    "            print('error')\n",
    "        #Based on the standard genetic code, codons code for a specific amino acid\n",
    "        #some codons also represent stopping points in translation, representing the \n",
    "        #end of translation, so these will break the loop to end the protein.\n",
    "        #Coded amino acids are added to the string 'protein'\n",
    "    return protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "364b79ae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Amino acid sequence of human hemoglobin beta:\n",
      "MVHLTPEEKSAVTALWGKVNVDEVGGEALGRLLVVYPWTQRFFESFGDLSTPDAVMGNPKVKAHGKKVLGAFSDGLAHLDNLKGTFATLSELHCDKLHVDPENFRLLGNVLVCVLAHHFGKEFTPPVQAAYQKVVAGVANALAHKYH\n"
     ]
    }
   ],
   "source": [
    "# Printing the Amino Acid Sequence of Human Hemoglobin Beta\n",
    "\n",
    "# The functions transcription and translation are carried out to convert the \n",
    "# gene to mRNA and then to the amino acid chain\n",
    "print(\"Amino acid sequence of human hemoglobin beta:\")\n",
    "print(translation(transcription(sequence)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "856a980d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
